File: Algorithms_medium/81_Search in Rotated Sorted Array II.py
class Solution:
    def search(self, nums: List[int], target: int) -> bool:
        l, r = 0, len(nums) - 1
        while l <= r:
            mid = (l + r) >> 1
            if nums[mid] == target:
                return True
            if nums[mid] < nums[r] or nums[mid] < nums[l]:
                if target > nums[mid] and target <= nums[r]:
                    l = mid + 1
                else:
                    r = mid - 1
            elif nums[mid] > nums[l] or nums[mid] > nums[r]:
                if target < nums[mid] and target >= nums[l]:
                    r = mid - 1
                else:
                    l = mid + 1
            else:
                r -= 1
        return False
'''
标准答案的二分查找，为啥要那么复杂我也没懂，我自己写了个遍历也过了，但感觉不对
'''

File: Algorithms_C#/81_Search_in_Rotated_Sorted_Array_II.cs
public class Solution {
    public bool Search(int[] nums, int target) {
        int l = 0, r = nums.Length - 1;
        while(l < r){
            int mid = (l + r) >> 1;
            if(nums[mid] > nums[r]){
                if(nums[l] <= target && target <= nums[mid]){
                    r = mid;
                }
                else{
                    l = mid + 1;
                }
            }
            else if(nums[mid] < nums[r]){
                if(nums[mid] < target && target <= nums[r]){
                    l = mid + 1;
                }
                else{
                    r = mid;
                }
            }
            else{
                --r;
            }          
        }
        return nums[l] == target;
    }
}